File: hooks/_changeset.mjs
// _changeset.mjs - каноническое множество изменений git-репозитория относительно
// базового коммита. Спецификация: skills/1c-code-review/references/changeset.md.
// Результат совпадает с tools/changeset.py байт в байт (сверяют тесты
// tests/tools/changeset/ по выводу CLI c --json). Экспорт computeChangeset
// предназначен хукам; CLI - тестам.
//
// CLI: node hooks/_changeset.mjs --repo <каталог> [--base <коммит>] [--json]
// Коды выхода: 0 расчет выполнен (в том числе пустое множество), 2 каталог не
// найден, не git-репозиторий, git недоступен, base не разрешается в коммит.

import { createHash } from 'node:crypto';
import { readFile, stat } from 'node:fs/promises';
import { execFile as execFileCb } from 'node:child_process';
import { resolve } from 'node:path';
import { promisify } from 'node:util';

const execFile = promisify(execFileCb);
const GIT_MAX_BUFFER = 64 * 1024 * 1024;

class ChangesetError extends Error {}

// Выполнить git и вернуть stdout (Buffer); любой отказ - ChangesetError с
// диагностикой git из stderr.
async function runGit(args, cwd) {
  try {
    const { stdout } = await execFile('git', args, { cwd, maxBuffer: GIT_MAX_BUFFER, encoding: 'buffer' });
    return stdout;
  } catch (err) {
    const stderr = Buffer.isBuffer(err.stderr) ? err.stderr.toString('utf8') : String(err.stderr || '');
    throw new ChangesetError(`git ${args.join(' ')} завершился с кодом ${err.code}: ${stderr.trim()}`);
  }
}

// Декодировать вывод git как UTF-8; невалидные байты (toString заменяет их на
// U+FFFD, round-trip меняет байты) - ChangesetError.
function decode(data, what) {
  const s = data.toString('utf8');
  if (Buffer.from(s, 'utf8').equals(data)) return s;
  throw new ChangesetError(`${what}: путь не является валидным UTF-8`);
}

// Разделить NUL-terminated вывод git на токены (без хвостового пустого).
function splitNul(buf) {
  const out = [];
  let start = 0;
  while (start < buf.length) {
    const idx = buf.indexOf(0, start);
    if (idx === -1) {
      out.push(buf.subarray(start));
      break;
    }
    out.push(buf.subarray(start, idx));
    start = idx + 1;
  }
  return out;
}

// Разобрать git diff --name-status -z в Map NFC-путь -> запись
// { status, renamedFrom, fsPath } по правилам спецификации. У R/C после кода идут
// ДВА пути: прежний, затем новый (не как в git status -z, где первым идет новый).
function parseDiff(raw) {
  const tokens = splitNul(raw);
  const entries = new Map();
  let i = 0;
  while (i < tokens.length) {
    const code = decode(tokens[i], 'git diff').trim();
    let renamedFrom = null;
    let fsPath;
    if (code.startsWith('R') || code.startsWith('C')) {
      renamedFrom = decode(tokens[i + 1], 'git diff').normalize('NFC');
      fsPath = decode(tokens[i + 2], 'git diff');
      i += 3;
    } else {
      fsPath = decode(tokens[i + 1], 'git diff');
      i += 2;
    }
    let status;
    if (code.startsWith('R')) status = 'renamed';
    else if (code.startsWith('A') || code.startsWith('C')) status = 'added';
    else if (code.startsWith('D')) status = 'deleted';
    else status = 'modified';
    entries.set(fsPath.normalize('NFC'), { status, renamedFrom, fsPath });
  }
  return entries;
}

// Отличается ли рабочий файл от base (с учетом фильтров и конверсии пути); для
// конфликта "файл есть в base и на диске, но не в индексе" (git rm --cached):
// git diff такой путь всегда видит удалением (untracked для него невидим), поэтому
// сравниваются blob-хеши: hash-object --path (фильтры как при add) против ls-tree.
// Путь-операнд отделен --: имя с ведущим дефисом иначе читается как ключ, значение
// --path передается присоединенным (--path=).
async function differsFromBase(top, base, fsPath) {
  const lsLine = decode(await runGit(['ls-tree', base, '--', fsPath], top), 'git ls-tree').trim();
  const fields = lsLine.split(/\s+/);
  if (fields.length < 3) return true; // пути нет в base; для конфликта с D недостижимо
  const baseBlob = fields[2].split('\t')[0];
  const workBlob = decode(
    await runGit(['hash-object', `--path=${fsPath}`, '--', fsPath], top), 'git hash-object').trim();
  return workBlob !== baseBlob;
}

// Пути-гитлинки (подмодули): режим 160000 в индексе или в базовом коммите. Gitlink -
// запись о коммите другого репозитория, в рабочем дереве это каталог; такой путь
// исключается из множества целиком при любом статусе diff. Источники - ls-files -s
// (индекс: новый gitlink в base отсутствует) и ls-tree -r <base> (gitlink мог быть
// удален из индекса). Формат -z: метаданные, TAB, путь, NUL - первый TAB отделяет
// метаданные от пути, TAB внутри самого пути не мешает.
async function gitlinkPaths(top, base) {
  const paths = new Set();
  const outputs = [await runGit(['ls-files', '-s', '-z'], top),
                   await runGit(['ls-tree', '-r', '-z', base], top)];
  for (const raw of outputs) {
    for (const record of splitNul(raw)) {
      const tabIdx = record.indexOf(9); // TAB
      if (tabIdx === -1) continue;
      const mode = record.subarray(0, tabIdx).toString('latin1').split(' ')[0];
      if (mode === '160000') {
        paths.add(decode(record.subarray(tabIdx + 1), 'git ls-tree').normalize('NFC'));
      }
    }
  }
  return paths;
}

// Собрать байты для diffHash из записей, уже отсортированных по байтам пути.
// Запись кадрируется однозначно: длина пути в байтах UTF-8 (десятичная), ":",
// байты пути, NUL, статус, NUL, sha256 либо литерал null, NUL. Префикс длины
// разделяет записи - имя с табуляцией или переводом строки не может сложиться
// в ту же последовательность байтов, что имена других файлов.
export function buildDiffPayload(files) {
  const chunks = [];
  for (const r of files) {
    const pathBytes = Buffer.from(r.path, 'utf8');
    const sha = r.sha256 === null ? 'null' : r.sha256;
    chunks.push(
      Buffer.from(`${pathBytes.length}:`, 'ascii'),
      pathBytes,
      Buffer.from(`\0${r.status}\0${sha}\0`, 'ascii'),
    );
  }
  return Buffer.concat(chunks);
}

// Вычислить каноническое множество изменений относительно base; результат по
// спецификации skills/1c-code-review/references/changeset.md:
// { base, diffHash, files } с записями { path, sha256, status [, renamedFrom] }.
export async function computeChangeset(repoDir, base = 'HEAD') {
  const repoPath = resolve(repoDir);
  let repoStat;
  try {
    repoStat = await stat(repoPath);
  } catch {
    throw new ChangesetError(`каталог не найден: ${repoPath}`);
  }
  if (!repoStat.isDirectory()) throw new ChangesetError(`не каталог: ${repoPath}`);

  const top = decode(await runGit(['rev-parse', '--show-toplevel'], repoPath), 'git rev-parse').trim();
  const baseSha = decode(
    await runGit(['rev-parse', '--verify', `${base}^{commit}`], top), 'git rev-parse').trim();
  const diffRaw = await runGit(
    ['diff', '--name-status', '-z', '-M', '--no-color', '--no-ext-diff', '--no-textconv', base], top);
  const untrackedRaw = await runGit(['ls-files', '--others', '--exclude-standard', '-z'], top);

  const entries = parseDiff(diffRaw);
  for (const token of splitNul(untrackedRaw)) {
    const fsPath = decode(token, 'git ls-files');
    const path = fsPath.normalize('NFC');
    if (!entries.has(path)) {
      entries.set(path, { status: 'added', renamedFrom: null, fsPath });
    }
    // Путь, совпавший с записью diff (git rm --cached без игнора), отдельной
    // записью не становится: статус из diff - deleted, арбитраж - в финальном
    // проходе по факту файла на диске.
  }

  const gitlinks = await gitlinkPaths(top, base);
  const files = [];
  for (const [path, entryIn] of entries) {
    let entry = entryIn;
    if (gitlinks.has(path)) continue; // подмодуль: gitlink, в рабочем дереве каталог
    let isFile = false;
    try {
      isFile = (await stat(resolve(top, entry.fsPath))).isFile();
    } catch {
      isFile = false;
    }
    if (entry.status === 'deleted' && isFile) {
      // Файл есть в base и на диске, но не в индексе (git rm --cached), причем
      // путь игнорируемый - ls-files --others его не показывает. Отличие от base
      // решает сравнение blob-хешей: равен - записи нет, отличается - modified.
      if (!(await differsFromBase(top, base, entry.fsPath))) continue;
      entry = { status: 'modified', renamedFrom: null, fsPath: entry.fsPath };
    }
    if (!isFile) {
      if (entry.status === 'added' || entry.status === 'renamed') continue; // нет ни в base, ни в дереве
      files.push({ path, sha256: null, status: 'deleted' });
      continue;
    }
    const buf = await readFile(resolve(top, entry.fsPath));
    const record = {
      path,
      sha256: createHash('sha256').update(buf).digest('hex'),
      status: entry.status,
    };
    if (entry.renamedFrom !== null) record.renamedFrom = entry.renamedFrom;
    files.push(record);
  }

  files.sort((a, b) => Buffer.compare(Buffer.from(a.path, 'utf8'), Buffer.from(b.path, 'utf8')));
  const diffHash = createHash('sha256').update(buildDiffPayload(files)).digest('hex');
  return { base: baseSha, diffHash, files };
}

// Рекурсивно отсортировать ключи объектов: JSON.stringify не сортирует сам,
// а сериализация обязана совпадать с json.dumps(sort_keys=True) из Python.
function sortKeysDeep(value) {
  if (Array.isArray(value)) return value.map(sortKeysDeep);
  if (value && typeof value === 'object') {
    const out = {};
    for (const key of Object.keys(value).sort()) out[key] = sortKeysDeep(value[key]);
    return out;
  }
  return value;
}

// Таблица множества для человека (формат спецификацией не фиксируется).
function renderTable(result) {
  const lines = [`base: ${result.base}`, `diffHash: ${result.diffHash}`, `файлов: ${result.files.length}`];
  for (const r of result.files) {
    const sha = r.sha256 === null ? '-' : r.sha256;
    const tail = 'renamedFrom' in r ? ` <- ${r.renamedFrom}` : '';
    lines.push(`${r.status.padEnd(8)} ${sha}  ${r.path}${tail}`);
  }
  return lines.join('\n');
}

async function cliMain() {
  const argv = process.argv.slice(2);
  const opts = { repo: '.', base: 'HEAD', json: false };
  for (let i = 0; i < argv.length; i++) {
    if (argv[i] === '--repo' && i + 1 < argv.length) opts.repo = argv[++i];
    else if (argv[i] === '--base' && i + 1 < argv.length) opts.base = argv[++i];
    else if (argv[i] === '--json') opts.json = true;
    else {
      process.stderr.write(`неизвестный аргумент: ${argv[i]}\n`);
      process.exit(2);
    }
  }
  try {
    const result = await computeChangeset(opts.repo, opts.base);
    if (opts.json) process.stdout.write(JSON.stringify(sortKeysDeep(result), null, 2) + '\n');
    else process.stdout.write(renderTable(result) + '\n');
    process.exit(0);
  } catch (err) {
    process.stderr.write(`ошибка: ${err.message}\n`);
    process.exit(2);
  }
}

// Запуск CLI только при прямом вызове (не при импорте тестами).
if (process.argv[1]?.endsWith('_changeset.mjs')) await cliMain();

File: tools/changeset.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Каноническое множество изменений git-репозитория относительно базового коммита.

Вычисляет записи изменений (путь, статус, sha256 рабочего файла), хеш множества
diffHash и полный SHA-1 базового коммита. Спецификация -
skills/1c-code-review/references/changeset.md; hooks/_changeset.mjs дает тот же
результат байт в байт (сверяется тестами tests/tools/changeset/).

Использование:
  python tools/changeset.py [--repo <каталог>] [--base <коммит>] [--json]

Без --json печатается таблица для человека.

Коды выхода: 0 расчет выполнен (в том числе пустое множество), 2 каталог не найден,
не git-репозиторий, git недоступен, base не разрешается в коммит.
"""

from __future__ import annotations

import argparse
import hashlib
import json
import subprocess
import sys
import unicodedata
from pathlib import Path

for _stream in (sys.stdout, sys.stderr):
    _reconf = getattr(_stream, "reconfigure", None)
    if callable(_reconf):
        try:
            # newline="\n" обязательно: без него на Windows перевод строки в пайпе
            # становится \r\n и вывод расходится с hooks/_changeset.mjs байт в байт.
            _reconf(encoding="utf-8", newline="\n")
        except (ValueError, OSError):
            pass


class ChangesetError(Exception):
    """Отказ расчета: git недоступен, каталог не репозиторий, base не разрешается, путь не UTF-8."""


def _run_git(args: list[str], cwd: Path) -> bytes:
    """Выполнить git с аргументами и вернуть stdout; любой отказ - ChangesetError.

    Диагностика git (stderr) попадает в сообщение исключения, вызывающий код
    переводит исключение в код выхода 2 без traceback.
    """
    try:
        proc = subprocess.run(["git", *args], cwd=str(cwd), capture_output=True)
    except OSError as exc:
        raise ChangesetError(f"git недоступен: {exc}") from exc
    if proc.returncode != 0:
        detail = proc.stderr.decode("utf-8", errors="replace").strip()
        raise ChangesetError(
            f"git {' '.join(args)} завершился с кодом {proc.returncode}: {detail}")
    return proc.stdout


def _decode(data: bytes, what: str) -> str:
    """Декодировать вывод git как UTF-8; невалидные байты - ChangesetError."""
    try:
        return data.decode("utf-8")
    except UnicodeDecodeError as exc:
        raise ChangesetError(f"{what}: путь не является валидным UTF-8") from exc


def _parse_diff(raw: bytes) -> dict[str, dict]:
    """Разобрать git diff --name-status -z в словарь NFC-путь -> запись.

    Запись: status, renamedFrom (None или прежний путь), fsPath (исходный путь из
    вывода git - по нему читается файл). Формат -z: поля записи разделены NUL -
    статус, путь; у R/C после статуса идут ДВА пути: прежний, затем новый
    (не как в git status -z, где первым идет новый).
    """
    tokens = raw.split(b"\0")
    entries: dict[str, dict] = {}
    i = 0
    while i < len(tokens) and tokens[i]:
        code = _decode(tokens[i], "git diff").strip()
        renamed_from = None
        if code.startswith(("R", "C")):
            renamed_from = unicodedata.normalize("NFC", _decode(tokens[i + 1], "git diff"))
            fs_path = _decode(tokens[i + 2], "git diff")
            i += 3
        else:
            fs_path = _decode(tokens[i + 1], "git diff")
            i += 2
        if code.startswith("R"):
            status = "renamed"
        elif code.startswith(("A", "C")):
            status = "added"
        elif code.startswith("D"):
            status = "deleted"
        else:
            status = "modified"
        entries[unicodedata.normalize("NFC", fs_path)] = {
            "status": status,
            "renamedFrom": renamed_from,
            "fsPath": fs_path,
        }
    return entries


def _differs_from_base(top: Path, base: str, fs_path: str) -> bool:
    """Отличается ли рабочий файл от base; решает git с учетом конверсий репозитория.

    Для конфликта "файл есть в base и на диске, но не в индексе" (git rm --cached):
    git diff такой путь всегда видит удалением (untracked для него невидим), поэтому
    отличие решается сравнением blob-хешей - git hash-object --path применяет к
    рабочим байтам фильтры и конверсию пути (как при add), git ls-tree base дает
    хеш блоба в базовом коммите. Путь-операнд отделен --: имя с ведущим дефисом
    иначе читается как ключ, значение --path передается присоединенным (--path=).
    """
    ls_line = _decode(_run_git(["ls-tree", base, "--", fs_path], top), "git ls-tree").strip()
    fields = ls_line.split()
    if len(fields) < 3:
        return True  # пути нет в base; для конфликта с D недостижимо, защита парсера
    base_blob = fields[2].split("\t")[0]
    work_blob = _decode(_run_git(["hash-object", f"--path={fs_path}", "--", fs_path], top),
                        "git hash-object").strip()
    return work_blob != base_blob


def _gitlink_paths(top: Path, base: str) -> set[str]:
    """Пути-гитлинки (подмодули): режим 160000 в индексе или в базовом коммите.

    Gitlink - запись о коммите другого репозитория, в рабочем дереве это каталог;
    спецификация исключает такой путь из множества целиком при любом статусе diff.
    Источники - git ls-files -s -z (индекс: новый gitlink в base отсутствует) и
    git ls-tree -r -z <base> (коммит: gitlink мог быть удален из индекса).
    """
    paths: set[str] = set()
    outputs = (_run_git(["ls-files", "-s", "-z"], top),
               _run_git(["ls-tree", "-r", "-z", base], top))
    for raw in outputs:
        for record in raw.split(b"\0"):
            if not record:
                continue
            # Формат -z у обеих команд: метаданные, TAB, путь, NUL. Первый TAB
            # отделяет метаданные от пути, TAB внутри самого пути не мешает.
            meta, _, path_bytes = record.partition(b"\t")
            if meta.split(b" ", 1)[0] == b"160000":
                paths.add(unicodedata.normalize("NFC", _decode(path_bytes, "git ls-tree")))
    return paths


def build_payload(files: list[dict]) -> bytes:
    """Собрать байты для diffHash из записей, уже отсортированных по байтам пути.

    Запись кадрируется однозначно: длина пути в байтах UTF-8 (десятичная), ":",
    байты пути, NUL, статус, NUL, sha256 либо литерал null, NUL. Префикс длины
    разделяет записи - имя с табуляцией или переводом строки не может сложиться
    в ту же последовательность байтов, что имена других файлов.
    """
    chunks: list[bytes] = []
    for record in files:
        path_bytes = record["path"].encode("utf-8")
        sha = record["sha256"] if record["sha256"] is not None else "null"
        chunks.append(f"{len(path_bytes)}:".encode("ascii") + path_bytes
                      + b"\0" + record["status"].encode("ascii")
                      + b"\0" + sha.encode("ascii") + b"\0")
    return b"".join(chunks)


def compute_changeset(repo_dir: Path | str, base: str = "HEAD") -> dict:
    """Вычислить каноническое множество изменений репозитория относительно base.

    Возвращает {"base": SHA-1, "diffHash": hex, "files": [записи]} по спецификации
    skills/1c-code-review/references/changeset.md. Базовые команды: rev-parse,
    diff --name-status, ls-files --others; untracked-пути сливаются с записями diff
    по правилам спецификации.
    """
    repo_dir = Path(repo_dir)
    if not repo_dir.is_dir():
        raise ChangesetError(f"каталог не найден: {repo_dir}")
    top = Path(_decode(_run_git(["rev-parse", "--show-toplevel"], repo_dir),
                       "git rev-parse").strip())
    base_sha = _decode(_run_git(["rev-parse", "--verify", f"{base}^{{commit}}"], top),
                       "git rev-parse").strip()
    diff_raw = _run_git(["diff", "--name-status", "-z", "-M",
                         "--no-color", "--no-ext-diff", "--no-textconv", base], top)
    untracked_raw = _run_git(["ls-files", "--others", "--exclude-standard", "-z"], top)

    entries = _parse_diff(diff_raw)
    for token in untracked_raw.split(b"\0"):
        if not token:
            continue
        fs_path = _decode(token, "git ls-files")
        path = unicodedata.normalize("NFC", fs_path)
        if path not in entries:
            entries[path] = {"status": "added", "renamedFrom": None, "fsPath": fs_path}
        # Путь, совпавший с записью diff (git rm --cached без игнора), отдельной
        # записью не становится: статус из diff - deleted, арбитраж - в финальном
        # проходе по факту файла на диске.

    gitlinks = _gitlink_paths(top, base)
    files: list[dict] = []
    for path, entry in entries.items():
        if path in gitlinks:
            continue  # подмодуль: gitlink, в рабочем дереве каталог, а не файл
        on_disk = (top / entry["fsPath"]).is_file()
        if entry["status"] == "deleted" and on_disk:
            # Файл есть в base и на диске, но не в индексе (git rm --cached), причем
            # путь игнорируемый - ls-files --others его не показывает. Отличие от
            # base решает сравнение blob-хешей: равен - записи нет, отличается -
            # modified.
            if not _differs_from_base(top, base, entry["fsPath"]):
                continue
            entry = {"status": "modified", "renamedFrom": None, "fsPath": entry["fsPath"]}
        if not on_disk:
            if entry["status"] in ("added", "renamed"):
                continue  # файла нет ни в base, ни в рабочем дереве
            files.append({"path": path, "sha256": None, "status": "deleted"})
            continue
        record = {
            "path": path,
            "sha256": hashlib.sha256((top / entry["fsPath"]).read_bytes()).hexdigest(),
            "status": entry["status"],
        }
        if entry["renamedFrom"] is not None:
            record["renamedFrom"] = entry["renamedFrom"]
        files.append(record)

    files.sort(key=lambda r: r["path"].encode("utf-8"))
    return {
        "base": base_sha,
        "diffHash": hashlib.sha256(build_payload(files)).hexdigest(),
        "files": files,
    }


def render_table(result: dict) -> str:
    """Таблица множества для человека: шапка с base и diffHash, строка на файл."""
    lines = [f"base: {result['base']}",
             f"diffHash: {result['diffHash']}",
             f"файлов: {len(result['files'])}"]
    for record in result["files"]:
        sha = record["sha256"] if record["sha256"] is not None else "-"
        tail = f" <- {record['renamedFrom']}" if "renamedFrom" in record else ""
        lines.append(f"{record['status']:<8} {sha}  {record['path']}{tail}")
    return "\n".join(lines)


def main(argv: list[str] | None = None) -> int:
    """Точка входа CLI: JSON (--json) либо таблица; 0 расчет выполнен, 2 отказ."""
    parser = argparse.ArgumentParser(
        description="Каноническое множество изменений git-репозитория относительно базового коммита.")
    parser.add_argument("--repo", type=Path, default=Path("."), metavar="КАТАЛОГ",
                        help="каталог репозитория (по умолчанию текущий)")
    parser.add_argument("--base", default="HEAD", metavar="КОММИТ",
                        help="базовый коммит (по умолчанию HEAD)")
    parser.add_argument("--json", action="store_true",
                        help="вывод JSON вместо таблицы")
    args = parser.parse_args(argv)
    try:
        result = compute_changeset(args.repo, args.base)
    except ChangesetError as exc:
        print(f"ошибка: {exc}", file=sys.stderr)
        return 2
    if args.json:
        print(json.dumps(result, ensure_ascii=False, indent=2, sort_keys=True))
    else:
        print(render_table(result))
    return 0


if __name__ == "__main__":
    sys.exit(main())
